internal/provider/kubernetes/controller.go

```diff
@@ -782,6 +769,47 @@ func (r *gatewayAPIReconciler) findReferenceGrant(ctx context.Context, from, to
 	return nil, nil
 }
 
+// processFinalizers encapsulates logic for managing finalizers on GatewayClass and EnvoyProxy objects.
+func (r *gatewayAPIReconciler) processFinalizers(ctx context.Context, managedGC *gwapiv1.GatewayClass, resourceTree *gatewayapi.Resources) error {
+	// Add finalizer to GatewayClass if there are gateways.
+	if len(resourceTree.Gateways) > 0 && managedGC.DeletionTimestamp.IsZero() {
+		if err := r.addFinalizer(ctx, managedGC); err != nil {
+			return fmt.Errorf("failed to add finalizer to gatewayclass %s: %w", managedGC.Name, err)
+		}
+	}
+
+	// Check previously stored referenced Envoy Proxy and remove finalizer if not referenced anymore by the current managed GatewayClass.
+	managedResources := r.resources.GetResourcesByGatewayClass(managedGC.Name)
+	if managedResources != nil {
+		refEnvoyProxy := managedResources.EnvoyProxyForGatewayClass
+		if refEnvoyProxy != nil && !classRefsEnvoyProxy(managedGC, refEnvoyProxy) {
+			if err := r.removeFinalizer(ctx, refEnvoyProxy); err != nil {
```

Comment:
EnvoyProxy can be referenced by multiple GatewayClasses.

Reply:
that's true, I have also tested this scenario, and `EnvoyProxy` finalizer will be there if it is referenced by another `GatewayClass`

Reply:
same thing happens here:

1. running `kubectl apply -f examples/kubernetes/merged-gateways.yaml` and create another GC (`foobar`) have envoyproxy `custom-proxy-config` as its `parametersRef`
2. `kubectl delete envoyproxy -n envoy-gateway-system custom-proxy-config` will hold
3. `kubectl delete gc merged-eg` will hold
4. `kubectl delete gtw merged-eg-1 merged-eg-2 merged-eg-3`, all gtws are removed successfully, but step 3 still hold (unexpected), surprisingly step 2 return (unexpected)

Reply:
getting back to this, let me take a look, it seems like we have another problem, now with getting errors within processParamsRef

Reply:
Thanks @shawnh2 @zhaohuabing
I needed to change the approach, because of the multiple gateway classes